# Pandas Learning Project 02: Police Datset

- Removing columns with missing values
- for speeding, where men & women stopped more often?
- Does the gender affect who gets searched during a stop
- Mean stop duration
- Age distribution for violations

In [2]:
import pandas as pd
import numpy as np

In [4]:
polics = pd.read_csv('polics.csv')
polics.head()

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   
2  1/23/2005     23:15           NaN             M          1972.0   
3  2/20/2005     17:15           NaN             M          1986.0   
4  3/14/2005     10:00           NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

In [5]:
polics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65535 entries, 0 to 65534
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           65535 non-null  object 
 1   stop_time           65535 non-null  object 
 2   country_name        0 non-null      float64
 3   driver_gender       61474 non-null  object 
 4   driver_age_raw      61481 non-null  float64
 5   driver_age          61228 non-null  float64
 6   driver_race         61475 non-null  object 
 7   violation_raw       61475 non-null  object 
 8   violation           61475 non-null  object 
 9   search_conducted    65535 non-null  bool   
 10  search_type         2479 non-null   object 
 11  stop_outcome        61475 non-null  object 
 12  is_arrested         61475 non-null  object 
 13  stop_duration       61475 non-null  object 
 14  drugs_related_stop  65535 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 6.6+

### Insights
---
Dataset has 15 columns & 65535 rows

---


# 1. Removing Column with missing values

In [6]:
polics.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

### Insights
---
It seems country name and search conducted column are totally blank

---


In [ ]:
# Removing a column using drop()

polics.drop(['country_name', 'search_type'], axis=1, inplace=True)

In [38]:
polics.head()

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0   1/2/2005      1:55             M          1985.0        20.0       White   
1  1/18/2005      8:15             M          1965.0        40.0       White   
2  1/23/2005     23:15             M          1972.0        33.0       White   
3  2/20/2005     17:15             M          1986.0        19.0       White   
4  3/14/2005     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted   stop_outcome is_arrested  \
0          Speeding  Speeding             False       Citation       False   
1          Speeding  Speeding             False       Citation       False   
2          Speeding  Speeding             False       Citation       False   
3  Call for Service     Other             False  Arrest Driver        True   
4          Speeding  Speeding             False       Citation       False   

  stop_duration  drugs_related_stop  mean_duration  
0      0-15 Min               False            7.5  
1      0-15 Min               False            7.5  
2      0-15 Min               False            7.5  
3     16-30 Min               False           23.0  
4      0-15 Min               False            7.5

# 2. For speeding, where Men & Women are stopped more often

In [25]:
speeding = polics[polics['violation'] == 'Speeding']
speeding['driver_gender'].value_counts()

M    25517
F    11686
Name: driver_gender, dtype: int64

### Insignts
---
Males are stopped 37% more often due to over speeding as compared to females. 

---


# 3. Does the gender affect who gets searched during a stop

In [31]:
# Filtering & counting values

search = polics[polics['search_conducted'] == True]
search['driver_gender'].value_counts()

M    2113
F     366
Name: driver_gender, dtype: int64

# 4. Mean stop duration

In [35]:
# Mapping in pandas

polics['mean_duration'] = polics['stop_duration'].map({'0-15 Min': 7.5, '16-30 Min': 23, '30+ Min': 15})

In [42]:
mean_value = polics['mean_duration'].mean()
print(f'Mean Stop Duration is {mean_value.round(2)}')

Mean Stop Duration is 10.71


# 5. Age distribution for violations

In [51]:
# Grouping Data

polics.groupby('violation').driver_age.describe()

count       mean        std   min   25%   50%   75%  \
violation                                                                    
Equipment             6507.0  31.682957  11.380671  16.0  23.0  28.0  39.0   
Moving violation     11876.0  36.736443  13.258350  15.0  25.0  35.0  47.0   
Other                 3477.0  40.362381  12.754423  16.0  30.0  41.0  50.0   
Registration/plates   2240.0  32.656696  11.150780  16.0  24.0  30.0  40.0   
Seat belt                3.0  30.333333  10.214369  23.0  24.5  26.0  34.0   
Speeding             37120.0  33.262581  12.615781  15.0  23.0  30.0  42.0   

                      max  
violation                  
Equipment            81.0  
Moving violation     86.0  
Other                86.0  
Registration/plates  74.0  
Seat belt            42.0  
Speeding             88.0